playback/src/audio_backend/alsa.rs

```diff
@@ -228,54 +236,96 @@ impl Sink for AlsaSink {
-            let data_to_buffer = min(
-                self.period_buffer.capacity() - self.period_buffer.len(),
-                data.len() - processed_data,
-            );
-            self.period_buffer
-                .extend_from_slice(&data[processed_data..processed_data + data_to_buffer]);
-            processed_data += data_to_buffer;
-            if self.period_buffer.len() == self.period_buffer.capacity() {
-                self.write_buf()?;
-                self.period_buffer.clear();
-            }
-        }
-
+        self.process_bytes(data, 0)?;
         Ok(())
     }
 }
 
 impl AlsaSink {
     pub const NAME: &'static str = "alsa";
 
-    fn write_buf(&mut self) -> io::Result<()> {
-        let pcm = self.pcm.as_mut().ok_or_else(|| {
-            io::Error::new(
-                io::ErrorKind::Other,
-                "Error writing from AlsaSink buffer to PCM, PCM is None",
-            )
-        })?;
-        let io = pcm.io_bytes();
-        if let Err(err) = io.writei(&self.period_buffer) {
-            // Capture and log the original error as a warning, and then try to recover.
-            // If recovery fails then forward that error back to player.
-            warn!(
-                "Error writing from AlsaSink buffer to PCM, trying to recover {}",
-                err
-            );
-            pcm.try_recover(err, false).map_err(|e| {
-                io::Error::new(
-                    io::ErrorKind::Other,
-                    format!(
-                        "Error writing from AlsaSink buffer to PCM, recovery failed {}",
-                        e
-                    ),
-                )
-            })?
+    fn configure_period_buffer(&mut self, bytes_per_period: usize) {
+        // Vec growth strategies can and have changed.
+        // Capacity may not always be trustworthy in the future.
+        // Use bytes_per_period to make sure that no matter what
+        // we give the PCM a period's worth of audio.
+        // Don't depend on len == capacity.
+        self.bytes_per_period = bytes_per_period;
+        // As noted above Vec growth strategies are not guaranteed
+        // so we over allocate initially so that our Vec will never
+        // actually reach full capacity thus (hopefully) avoiding
+        // future reallocations. Nearest (truncated) KiB + 2 KiB.
+        let over_allocation = ((bytes_per_period / KIB) * KIB) + TWO_KIB;
+        // If the capacity is greater than we want try to shrink it
+        // to it's current len (which should be zero) before trying
+        // to set the capacity with reserve_exact.
+        if self.period_buffer.capacity() > over_allocation {
+            self.period_buffer.shrink_to_fit();
+        }
+        // This does nothing if the capacity is already sufficient.
+        // Len should always be zero, but for the sake of being thorough...
+        self.period_buffer
+            .reserve_exact(over_allocation - self.period_buffer.len());
```

Comment:
So after the "`Vec` deep dive" it seems that `Vec::with_capacity(x)` guarantees `x` but `reserve` and `reserve_exact` don't. In that case you could just do `self.period_buffer = Vec::with_capacity(x)` and be done with it.

Reply:
And now you believe me,lol!!! Ok. Well that was basically the whole reason for most of these changes... You're worse than my wife. Have you made up your mind for real this time?

Reply:
Good thing we're not getting married! 😆 Like I said proof of the pudding is in the eating and your gist did show `with_capacity` is a friend of ours. So yeah this is the dress I'm going to wear to the dance.

Reply:
It also showed that `reserve_exact` is our friend also and can be trusted so that means w could still reuse the buffer.

Reply:
Done. Not a thing anymore.